src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -504,20 +504,42 @@ memory_snapshot_harness_generatort::initialize_entry_via_source(
 {
   PRECONDITION(!entry_source_location.file_name.empty());
 
+  size_t best_match_distance = -1;
+  irep_idt best_match_function_name;
+  std::tuple<goto_programt::const_targett, irep_idt, size_t> best_match;
+
   // by line: iterate over all instructions until source location match
   for(const auto &entry : goto_functions.function_map)
   {
     const auto &goto_function = entry.second;
     // if !body_available() then body.instruction.empty() and that's fine
     const auto &goto_program = goto_function.body;
 
-    const auto corresponding_instruction =
+    const auto candidate_instruction =
       entry_source_location.find_first_corresponding_instruction(
         goto_program.instructions);
 
-    if(corresponding_instruction != goto_program.instructions.end())
-      return entry_locationt{entry.first, corresponding_instruction};
+    if(candidate_instruction.first != goto_program.instructions.end())
+    {
+      if(best_match_distance > candidate_instruction.second)
+      {
+        best_match_function_name = entry.first;
+        best_match_distance = candidate_instruction.second;
```

Comment:
I think we can record `candidate_instruction.first` here as well, then we don't need to look it up again in the code block below.

Reply:
I wrapped those variables in a struct and added the instruction as well.